src/utils/multipart-parser.js

```diff
@@ -0,0 +1,440 @@
+				case S.HEADER_VALUE:
+					if (c === CR) {
+						dataCallback('headerValue', true);
+						callback('headerEnd');
+						state = S.HEADER_VALUE_ALMOST_DONE;
+					}
+
+					break;
+				case S.HEADER_VALUE_ALMOST_DONE:
+					if (c !== LF) {
+						return i;
+					}
+
+					state = S.HEADER_FIELD_START;
+					break;
+				case S.HEADERS_ALMOST_DONE:
+					if (c !== LF) {
+						return i;
+					}
+
+					callback('headersEnd');
+					state = S.PART_DATA_START;
+					break;
+				case S.PART_DATA_START:
+					state = S.PART_DATA;
+					mark('partData');
+					// fallsthrough
+				case S.PART_DATA:
+					previousIndex = index;
+
+					if (index === 0) {
+						// boyer-moore derrived algorithm to safely skip non-boundary data
+						i += boundaryEnd;
+						while (i < bufferLength && !(ui8a[i] in boundaryChars)) {
+							i += boundaryLength;
+						}
+
+						i -= boundaryEnd;
+						c = ui8a[i];
+					}
+
+					if (index < boundary.length) {
+						if (boundary[index] === c) {
+							if (index === 0) {
+								dataCallback('partData', true);
+							}
+
+							index++;
+						} else {
+							index = 0;
+						}
+					} else if (index === boundary.length) {
+						index++;
+						if (c === CR) {
+							// CR = part boundary
+							flags |= F.PART_BOUNDARY;
+						} else if (c === HYPHEN) {
+							// HYPHEN = end boundary
+							flags |= F.LAST_BOUNDARY;
+						} else {
+							index = 0;
+						}
+					} else if (index - 1 === boundary.length) {
+						if (flags & F.PART_BOUNDARY) {
+							index = 0;
+							if (c === LF) {
+								// unset the PART_BOUNDARY flag
+								flags &= ~F.PART_BOUNDARY;
+								callback('partEnd');
+								callback('partBegin');
+								state = S.HEADER_FIELD_START;
+								break;
+							}
+						} else if (flags & F.LAST_BOUNDARY) {
+							if (c === HYPHEN) {
+								callback('partEnd');
+								callback('end');
+								state = S.END;
+								flags = 0;
+							} else {
+								index = 0;
+							}
+						} else {
+							index = 0;
+						}
+					}
+
+					if (index > 0) {
+						// when matching a possible boundary, keep a lookbehind reference
+						// in case it turns out to be a false lead
+						lookbehind[index - 1] = c;
+					} else if (previousIndex > 0) {
+						// if our boundary turned out to be rubbish, the captured lookbehind
+						// belongs to partData
+						const _lookbehind = new Uint8Array(lookbehind.buffer, lookbehind.byteOffset, lookbehind.byteLength);
+						callback('partData', 0, previousIndex, _lookbehind);
+						previousIndex = 0;
+						mark('partData');
+
+						// reconsider the current character even so it interrupted the sequence
+						// it could be the beginning of a new sequence
+						i--;
+					}
+
+					break;
+				case S.END:
+					break;
+				default:
+					return i;
+			}
+		}
+
+		dataCallback('headerField');
+		dataCallback('headerValue');
+		dataCallback('partData');
+
+		this.index = index;
+		this.state = state;
+		this.flags = flags;
```

Comment:
What is the point of extracting these variables and then assigning them back? 🤔 

Is it to avoid exposing some kind of in-between state outwards? Maybe we should add a comment about it?

Reply:
I think this was a speed optimization provided by the original author.

Before entering the the hole for-loop and doing lots of chain lookup `this.state = x` then i guess it was faster to just do `state = x` and then update the property later.

Reply:
@LinusU yea, something like that. Every time i dig a bit deeper in Formidable's source i'm shocked what a parser that is. But well, it's here for 10 years and is working. Felix is.. ah.